File: modules/07b-delivery-operations/py/m07b_service/provider.py
"""The ``llm_core`` seam.

Every model call in this service goes through the provider-agnostic client — we
never import an OpenAI/Anthropic SDK directly. Tests inject a fake provider that
implements the same :class:`~llm_core.LLMProvider` protocol, so the request path
is exercised deterministically and offline.
"""

from __future__ import annotations

from llm_core import LLMProvider, get_provider

from .config import Settings


def build_default_provider(settings: Settings) -> LLMProvider:
    """Construct the real provider selected by configuration.

    ``get_provider`` reads the provider-specific credentials from the
    environment itself; we only pass the validated provider name.
    """
    return get_provider(settings.provider)

File: modules/07b-delivery-operations/ts/src/provider.ts
/**
 * The `@learn-ai/llm-core` seam.
 *
 * Every model call goes through the provider-agnostic client — we never import a
 * vendor SDK directly. Tests inject a fake provider implementing the same
 * `LLMProvider` interface, so the request path runs deterministically offline.
 */

import type { Config } from "./config.js";
import type { LLMProvider, ProviderName } from "@learn-ai/llm-core";

/**
 * Construct the real provider selected by configuration.
 *
 * `@learn-ai/llm-core` is imported *dynamically* (not at module top level) on
 * purpose: it is an ESM-only package that resolves env/credentials at import
 * time. Loading it lazily keeps it out of the synchronous module graph, so tests
 * that inject a fake provider never load the real client, and credential
 * resolution is deferred to the first real model call.
 */
export async function buildDefaultProvider(config: Config): Promise<LLMProvider> {
  const { getProvider } = await import("@learn-ai/llm-core");
  // `getProvider` reads the provider-specific credentials from the environment;
  // we only pass the validated provider name.
  return getProvider(config.provider as ProviderName);
}
